bulk_indexer.go

```diff
@@ -257,52 +271,82 @@ func (b *bulkIndexer) Flush(ctx context.Context) (BulkIndexerResponseStat, error
 		return resp, fmt.Errorf("error decoding bulk response: %w", err)
 	}
 
-	buf := make([]byte, 1024)
+	if b.maxDocumentRetry != 0 {
+		buf := make([]byte, 1024)
+
+		for k := range b.retryCounts {
+			found := false
+			for _, res := range resp.FailedDocs {
+				if res.Position == k {
+					found = true
+				}
+			}
+			if !found {
+				delete(b.retryCounts, k)
+			}
+		}
 
-	for _, res := range resp.FailedDocs {
-		if res.Status == http.StatusTooManyRequests {
-			startlnIdx := res.Position * 2
-			endlnIdx := startlnIdx + 2
+		tmp := resp.FailedDocs[:0]
+		for _, res := range resp.FailedDocs {
+			if res.Status == http.StatusTooManyRequests {
+				startlnIdx := res.Position * 2
+				endlnIdx := startlnIdx + 2
 
-			if b.gzipw != nil {
-				gr, err := gzip.NewReader(bytes.NewReader(copyBuf))
-				if err != nil {
-					return resp, fmt.Errorf("failed to decompress request payload: %w", err)
+				count := b.retryCounts[res.Position] + 1
+				if count > b.maxDocumentRetry {
+					tmp = append(tmp, res)
+					continue
 				}
-				defer gr.Close()
 
-				seen := 0
+				b.retryCounts[b.itemsAdded] = count
```

Comment:
I think a couple of comments would go a long way:
```go
				// Increment 429 count for the positions found.
				count := b.retryCounts[res.Position] + 1
				if count > b.maxDocumentRetry {
					tmp = append(tmp, res)
					continue
				}
				// Since some items may have succeeded, counter positions need
				// to be updated to match the next current buffer position.
				b.retryCounts[b.itemsAdded] = count
```

I like the retry tracking algorithm, it's neat!